### PR TITLE
Mutating Interface Compute Tags and Performance Improvements

### PR DESCRIPTION
**Pull request by nilsdeppe**
## Proposed changes

- Allow interface compute tags to be mutating
- Reduce allocations in GRMHD cons from prim

Commit order:
1. Have compute interface tags wrapper mutate map
2. Make db::has_member_return_type available
3. Add support for mutating interface compute items
4. Add and use not_null variant of data_on_slice
5. Make Valencia Div Clean characteristic_speeds not_null
6. Reduce allocations in GRMHD ConservativeFromPrimitive

Depends on:
- [x] #1160 
- [x] #1137 

### Types of changes:

- [ ] Bugfix
- [x] New feature

### Component:

- [x] Code
- [ ] Documentation
- [ ] Build system
- [ ] Continuous integration

### Code review checklist

- [ ] The PR passes all checks, including unit tests, `clang-tidy` and `IWYU`. For
  instructions on how to perform the CI checks locally refer to the [Dev guide
  on the Travis CI](https://spectre-code.org/travis_guide.html).
- [ ] The code is documented and the documentation renders correctly. Run `make doc`
  to generate the documentation locally into `BUILD_DIR/docs/html`. Then open
  `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).

### CODE REVIEW

**Comment by nilsdeppe** [comment]
Rebased, and I removed the dependency.

**Comment by nilsvu** [comment]
are these all non-breaking changes? just making sure

**Comment by nilsdeppe** [comment]
I think so. If you rely on things in a `detail` namespace you're at fault, and if you rely on interface items using unordered_map, you're also at fault. Both are implementation details that should not be relied upon. This goes back to the problem that in some sense every change could be breaking for somebody.

**Comment by nilsvu** [comment]
both cases you mentioned I'd consider reasonably non-breaking. thanks 👍

File: src/Evolution/Systems/GrMhd/ValenciaDivClean/ConservativeFromPrimitive.cpp [hunk 1]
```diff
@@ -39,17 +39,47 @@ void ConservativeFromPrimitive::apply(
...
+      hydro::Tags::SpatialVelocityOneForm<DataVector, 3, Frame::Inertial>,
+      hydro::Tags::SpatialVelocitySquared<DataVector>,
+      hydro::Tags::MagneticFieldOneForm<DataVector, 3, Frame::Inertial>,
+      hydro::Tags::MagneticFieldDotSpatialVelocity<DataVector>,
+      hydro::Tags::MagneticFieldSquared<DataVector>>>
+      temp_tensors{get(rest_mass_density).size()};
+  const auto& spatial_velocity_one_form =
+      get<hydro::Tags::SpatialVelocityOneForm<DataVector, 3, Frame::Inertial>>(
+          temp_tensors);
+  raise_or_lower_index(
```
Comment (kidder) [inline src/Evolution/Systems/GrMhd/ValenciaDivClean/ConservativeFromPrimitive.cpp]: why not mutating references that can be used in the `make_not_null` calls?

Comment (nilsdeppe) [reply src/Evolution/Systems/GrMhd/ValenciaDivClean/ConservativeFromPrimitive.cpp]: Sorry, I was previously doing this so we would have more const things but people seem to prefer mutating references from code review, so I will update.

Comment (nilsdeppe) [reply src/Evolution/Systems/GrMhd/ValenciaDivClean/ConservativeFromPrimitive.cpp]: Done and rebased on latest develop

**Comment by kidder** [comment]
@wthrowe @markscheel @fmahebert please look at

**Comment by fmahebert** [comment]
Can't speak to the DataBox commits, but the Valencia/GRMHD commits look good to me :+1:

**Review by kidder — DISMISSED** [review]

File: src/Domain/Tags.hpp [hunk 2]
```diff
@@ -252,15 +252,21 @@ struct unmap_interface_args;
...
 template <>
 struct unmap_interface_args<true> {
+  template <typename T>
+  using f = T;
+
   template <size_t VolumeDim, typename T>
-  static constexpr decltype(auto) apply(
-      const ::Direction<VolumeDim>& /*direction*/, const T& arg) noexcept {
+  static constexpr const T& apply(const ::Direction<VolumeDim>& /*direction*/,
+                                  const T& arg) noexcept {
```
Comment (wthrowe) [inline src/Domain/Tags.hpp:260]: What's this change about?  Didn't the return type work out to `const T&` already?

File: src/DataStructures/DataBox/DataBox.hpp [hunk 3]
```diff
@@ -684,9 +684,9 @@ struct compute_item_function_pointer_type_impl<true> {
 template <typename FullTagList, typename ComputeItem,
```
Comment (wthrowe) [inline src/DataStructures/DataBox/DataBox.hpp:684]: Commit message uses wrong alias name.

File: src/Evolution/Systems/GrMhd/ValenciaDivClean/Characteristics.hpp [hunk 4]
```diff
@@ -112,6 +113,22 @@ std::array<DataVector, 9> characteristic_speeds(
     const EquationsOfState::EquationOfState<true, ThermodynamicDim>&
         equation_of_state) noexcept;
 
+template <size_t ThermodynamicDim>
+void characteristic_speeds(
```
Comment (wthrowe) [inline src/Evolution/Systems/GrMhd/ValenciaDivClean/Characteristics.hpp:117]: I'm confused.  Am I reading this correctly that this function already existed but was missing a declaration in the header file?

**Review by wthrowe — CHANGES_REQUESTED** [review]

File: src/Evolution/Systems/GrMhd/ValenciaDivClean/Characteristics.hpp (hunk 4 above)
Comment (nilsdeppe) [reply src/Evolution/Systems/GrMhd/ValenciaDivClean/Characteristics.hpp:117]: Yes, before it was only used internally from the not not_null version.

File: src/Domain/Tags.hpp (hunk 2 above)
Comment (nilsdeppe) [reply src/Domain/Tags.hpp:260]: I find it easier if in these simple cases the type is more explicit and was useful in debugging the template issues to convince myself the code was behaving as expected :)

File: src/DataStructures/DataBox/DataBox.hpp (hunk 3 above)
Comment (nilsdeppe) [reply src/DataStructures/DataBox/DataBox.hpp:684]: Okay, reworded :)

**Review by wthrowe — APPROVED** [review]

**Review by kidder — APPROVED** [review]

**Pull request merged**